sigs/jvm/rfcs/20200126-tensorflow-data.md

```diff
@@ -0,0 +1,209 @@
+load, transform, and iterate over datasets.
+
+Additionally this framework will be used at the core of the automated
+training process in TensorFlow Java's port of [tf.keras](https://www.tensorflow.org/api_docs/python/tf/keras).
+
+## Design Proposal
+
+The `Dataset` class represents a sequence of elements which can be iterated over and
+transformed. Each element is a list of "output" operands, represented by the type `List<Output<?>>`. 
+
+Note: An `Output` is a symbolic handle to a tensor produced by a TensorFlow op. In graph
+mode, `Output` objects will not have a concrete `Tensor` value unless all dependent operations
+are run in a `Session` (this is done "in-real-time" in eager mode).
+
+### Construction
+
+Datasets can be constructed either directly from a data source (e.g. a list of tensors representing the components of the dataset), or as a transformation on an existing dataset.
+
+#### From Data Source
+
+To construct a dataset from a list of tensor components, use 
+`Dataset.fromTensorSlices( ... )`. For example, say we are working
+with a standard feature/label dataset which has 4 elements.
+
+```java
+float[][] features = new float[][] {
+    {1, 2, 3},
+    {4, 5, 6},
+    {7, 8, 9},
+    {10, 11, 12}
+}
+
+float[] labels = new float[] {
+    0,
+    1,
+    1,
+    0
+}
+```
+
+A dataset can be constructed from a list of the constant `Operand`s generated
+from this dataset, and a list of `DataType` objects corresponding
+to the type of each component:
+
+Note: Each of the input components must share the same first "batch" dimension.
+
+```java
+Ops tf = // ... TensorFlow Ops accessor (either graph or eager)
+Dataset dataset = Dataset.fromTensorSlices(
+    Arrays.asList(tf.constant(features), tf.constant(labels)),
+    Arrays.asList(TInt32.DTYPE, TInt32.DTYPE)
+);
+```
+
+
+Other data sources are also possible, using `tf.data` ops. Datasets can also be constructed from 
+
+#### Transformations on Existing Datasets
+
+Once a dataset has been created from a data source it can be transformed by calling
+methods on the `Dataset` object. For example, to group elements in the above dataset into batches of size `2`, use `Dataset.batch(int batchSize)`:
+
+```java
+dataset = dataset.batch(2)
+```
+
+Dataset transformations alter both the values and shapes of the original elements, and
+return a *new* `Dataset` object.
+
+In this case, the original dataset had 4 elements of shape `[features: (3,) labels: (1,)]`.
+Once the `.batch` transformation is applied, the new dataset has 2 elements (batches) of shape `[features: (2, 3), labels: (2, 1)]`.
+
+Similar transformations include `.skip`, `.take`, `.map`, `.filter`, etc.
+
+
+### Iterating over Dataset Elements
+
+The primary use of a dataset is for iteration over its elements.
+Each row (or batch) element is represented as a list of tensor components, with
+type `List<Output<?>>`. The tensor components of this elements can be accessed using `List.get(int index)`.
```

Comment:
`List.get(int)` will return you a tensor, not an output, no? 

Maybe it would be interesting to leave the batch items as `Operand` or `Constant` objects? `Output` became in time kind of a more internal data class.

Reply:
This method could return `List<Operand<?>>` (can't return a `Constant`, I think, since these are outputs of tf.data ops).

I think users won't often have to work with `Tensor` very often here, since these data values only need to be passed to other operations, in most workflows.

Reply:
We could change this to be a `List<Operand<?>>`, if we modify the core `tf.data.iteratorGetNextOp`.